File: supermarket-pricing/typescript/src/Product.ts
import type { PricingRule } from './PricingRule.js';

export class Product {
  constructor(
    public readonly sku: string,
    public readonly name: string,
    public readonly pricingRule: PricingRule,
  ) {
    if (sku.trim().length === 0) {
      throw new Error('SKU must not be empty');
    }
  }
}

File: supermarket-pricing/python/src/supermarket_pricing/product.py
from __future__ import annotations

from dataclasses import dataclass

from .pricing_rule import PricingRule


@dataclass(frozen=True)
class Product:
    sku: str
    name: str
    pricing_rule: PricingRule

    def __post_init__(self) -> None:
        if not self.sku or not self.sku.strip():
            raise ValueError("SKU must not be empty")
